addons/website/tests/test_redirect.py

```diff
@@ -67,3 +67,170 @@ def test_sitemap_with_redirect(self):
             self.assertIn('/website/info', urls)
             self.assertNotIn('/test', urls)
+
+
+@tagged('-at_install', 'post_install')
+class TestConditionalRedirects(HttpCase):
+
+    def setUp(self):
+        super().setUp()
+        self.test_group = self.env['res.groups'].create({'name': 'Test Group'})
+        # User who is a member of the test group
+        self.user_in_group = self.env['res.users'].with_context({'no_reset_password': True}).create({
+            'name': 'Test User In Group',
+            'login': 'test_user_in_group',
+            'password': 'test_user_in_group',
+            'email': 'test_user_in_group@mail.com',
+            'group_ids': [(6, 0, [self.env.ref('base.group_portal').id, self.test_group.id])]
+        })
+        # User who is NOT a member of the test group
+        self.user_not_in_group = self.env['res.users'].with_context({'no_reset_password': True}).create({
+            'name': 'Test User Not In Group',
+            'login': 'test_user_not_in_group',
+            'password': 'test_user_not_in_group',
+            'email': 'test_user_not_in_group@mail.com',
+            'group_ids': [(6, 0, [self.env.ref('base.group_portal').id])]
+        })
+        self.original_page = self.env['website.page'].create({
+            'name': 'Original Page',
+            'url': '/original-page',
+            'is_published': False,
+            'type': 'qweb',
+            'arch': '<div>Original Page Content</div>',
+        })
+        self.redirected_page = self.env['website.page'].create({
+            'name': 'Redirected Page',
+            'url': '/redirected-page',
+            'is_published': True,
+            'type': 'qweb',
+            'arch': '<div>Redirected Page Content</div>',
+        })
+        self.url_from_301 = '/original-page?redirect=301'
+        self.url_to_301 = '/redirected-page?redirect=301'
+        self.url_from_302 = '/original-page?redirect=302'
+        self.url_to_302 = '/redirected-page?redirect=302'
+        self.url_from_404 = '/website/info'
+
+    def _create_rewrite_rules(self, apply_to, redirect_types=None):
+        """Helper to create a set of rewrite rules for testing."""
+        if redirect_types is None:
+            redirect_types = ['301', '302', '404']
+        for r_type in redirect_types:
+            url_from = f"/original-page?redirect={r_type}" if r_type != '404' else self.url_from_404
+            url_to = f"/redirected-page?redirect={r_type}" if r_type != '404' else None
+            vals = {
+                'name': f'Test {apply_to} {r_type}',
+                'redirect_type': r_type,
+                'url_from': url_from,
+                'apply_to_group': apply_to,
+                'user_group_ids': [(6, 0, [self.test_group.id])] if apply_to != 'all_users' else False,
+            }
+            if url_to:
+                vals['url_to'] = url_to
+            self.env['website.rewrite'].create(vals)
+
+    def test_01_all_users(self):
```

Comment:
And looking at all the test I see a lot of repetitive code and makes it difficult to understand, maybe we can simplify it.

* Make a helper for to test redirect behavior maybe `_test_redirect_behavior` where you can authenticate user and validate expected redirect behavior. For all cases.